### ISSUE TITLE
Documentation Clarification on lists

### ISSUE THREAD

**Issue by ThinkDigitalSoftware** [hivedb/hive#141]
According to the ["Modelling" Data docs](https://docs.hivedb.dev/best-practices/modelling_data#lists-vs-auto-increment), it says: 
> If you want to store a list of items, you have two options. Either store the list directly as a value (using put('myKey', [1, 2, 3])) or you can store each item individually (using add(1), add(2) etc.).
>
> If you want to store a big list which is updated frequently (for example a list of messages), you should store the items individually. Otherwise the whole list has to be written each time you change an item.

Could you indicate the structure of the data for each of these operations? I understand the first option to have a structure like so.
```json
{'myKey': [1, 2, 3]}
```
but not what the structure of the second would be.

**Comment by leisim** [comment]
I agree, this needs improvement.

> I understand the first option to have a structure like so.
> `{'myKey': [1, 2, 3]}`

Correct.

Let's say you do `add('item1')`, `add('item2')`, `add('item3')`.

The structure would be:
```json
{
  0: 'item1',
  1: 'item2',
  2: 'item3' 
}
```

**Comment by ThinkDigitalSoftware** [comment]
Ok, I didn't expect that. Especially since it's a key value DB, having an add command without specifying a key means we have to look it up without knowing what it is. What would happen if we add key 2 using put, and then add 3 listitems? How would we know how to retrieve that specific list? Because option one isn't the same as option 2. The way the docs read, option 2 is a more efficient version one option one for large lists, but they don't serve the same purpose

**Comment by thedumbtechguy** [comment]
This actually interested me. Glad to see it raised. 
Personally, I've decided to go with `put` exclusively and setting my own keys.
Not sure if it does that, but if `add` returns the key it set, it should be cool. Else, I think it will mostly lead to problems.

**Comment by ThinkDigitalSoftware** [comment]
Agreed

**Comment by leisim** [comment]
You can use a box either like a map using `put()` and keys or like a list using `add()` and indices.

> Else, I think it will mostly lead to problems.

If you add an object to a list it doesn't return the index and you still know it ;)

**Comment by ThinkDigitalSoftware** [comment]
> You can use a box either like a map using `put()` and keys or like a list using `add()` and indices.
> 
> > Not sure if it does that, but if add returns the key it set, it should be cool. Else, I think it will mostly lead to problems.
> 
> If you add an object to a list it doesn't return the index and you still know it ;)

But key value stores mirror maps more Than they mirror lists. Which would just be a value store.

**Comment by leisim** [comment]
> But key value stores mirror maps more Than they mirror lists.

I agree but it will be very useful once queries are supported. I'm not sure how else to do it...

**Comment by ThinkDigitalSoftware** [comment]
I can't wait for queries. I was looking at switching to moor because it supports it, but I'd stick around if we had that :)

**Comment by leisim** [comment]
First I'll implement backlinks and queries after that. Moor is also a great library. You can also use the two together...

**Comment by ThinkDigitalSoftware** [comment]
Hmm. I can't see a use case for using both yet.

**Comment by ThinkDigitalSoftware** [comment]
Also, what are back links?

**Comment by leisim** [comment]
> Also, what are back links?

For example if you want to store a user class:
```dart
class User extends HiveObject {
  String name;

  List<User> friends; // <- backlink
}
```

The friends are users themselfs and the list only contains "links". You can also link to objects of another box.

**Comment by ThinkDigitalSoftware** [comment]
So like Cloud Firestore's document references?

**Comment by leisim** [comment]
> I can't see a use case for using both yet.

You could for example store relational data in SQLite and preferences, performance critical stuff etc. in Hive.

> So like Cloud Firestore's document references?

I don't use Firestore much but it sounds like a similar concept.
